File: gremlin-python/src/main/python/gremlin_python/driver/connection.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
import asyncio
import queue
from concurrent.futures import Future

from aiohttp.client_exceptions import (
    ClientOSError,
    ClientPayloadError,
    ServerDisconnectedError,
)

from gremlin_python.driver import resultset, useragent
from gremlin_python.driver.aiohttp.transport import AiohttpHTTPTransport
from gremlin_python.driver.http_request import HttpRequest

__author__ = 'David M. Brown (davebshow@gmail.com)'

_TRANSPORT_ERRORS = (ClientOSError, ClientPayloadError, ServerDisconnectedError, asyncio.IncompleteReadError)
_CONNECTION_ERROR_MSG = (
    "Connection to server closed unexpectedly. "
    "Ensure that the server is still reachable and the connection has not been closed by the server or a network device."
)


class GremlinServerError(Exception):
    def __init__(self, status):
        super(GremlinServerError, self).__init__('{0}: {1}'.format(status['code'], status['message']))
        self.status_code = status['code']
        self.status_message = status['message']
        self.status_exception = status['exception']


class GremlinConnectionError(Exception):
    """Raised when a transport-level failure occurs communicating with the server."""
    pass


class Connection:

    def __init__(self, url, traversal_source,
                 executor, pool,
                 response_serializer=None, auth=None, interceptors=None,
                 enable_user_agent_on_connect=True,
                 bulk_results=False, **transport_kwargs):
        if callable(interceptors):
            interceptors = [interceptors]
        elif isinstance(interceptors, tuple):
            interceptors = list(interceptors)
        elif not (isinstance(interceptors, list) or interceptors is None):
            raise TypeError("interceptors must be a callable, tuple, list or None")

        # Auth is just an interceptor. As a convenience (and for discoverability), the auth
        # interceptor is appended to the end of the interceptor list so it runs last, after
        # any user interceptors have modified the request.
        if auth is not None:
            interceptors = (interceptors or []) + [auth]

        self._url = url
        # Custom request headers are set via interceptors. This internal dict
        # only carries connection-level headers managed by the driver itself
        # (user agent, bulkResults) and is established at construction time.
        self._headers = None
        self._traversal_source = traversal_source
        self._transport_kwargs = transport_kwargs
        self._executor = executor
        self._transport = None
        self._pool = pool
        self._result_set = None
        self._inited = False
        self._response_serializer = response_serializer
        self._interceptors = interceptors
        self._enable_user_agent_on_connect = enable_user_agent_on_connect
        if self._enable_user_agent_on_connect:
            self.__add_header(useragent.userAgentHeader, useragent.userAgent)
        self._bulk_results = bulk_results
        if self._bulk_results:
            self.__add_header("bulkResults", "true")

    def connect(self):
        if self._transport:
            self._transport.close()
        self._transport = AiohttpHTTPTransport(**self._transport_kwargs)
        self._transport.connect(self._url, self._headers)
        self._inited = True

    def close(self):
        if self._inited:
            self._transport.close()

    def _write_request(self, request_message):
        accept = str(self._response_serializer.version, encoding='utf-8')

        headers = {'accept': accept}

        # Promote transactionId to HTTP header before interceptors run.
        # The field remains in the serialized body as well (dual transmission
        # per the HTTP transaction protocol specification).
        if hasattr(request_message, 'fields') and 'transactionId' in request_message.fields:
            headers['X-Transaction-Id'] = request_message.fields['transactionId']

        http_request = HttpRequest(
            method="POST",
            url=self._url,
            headers=headers,
            body=request_message
        )

        for interceptor in self._interceptors or []:
            interceptor(http_request)

        # Auto-serialize if no interceptor already did so
        http_request.serialize_body()

        # Build the transport message in the format the transport expects
        message = {
            'headers': http_request.headers,
            'payload': http_request.body
        }
        self._transport.write(message)

    def write(self, request_message):
        if not self._inited:
            self.connect()
        self._result_set = resultset.ResultSet(queue.Queue())
        # Create write task
        future = Future()
        future_write = self._executor.submit(
            self._write_request, request_message)

        def cb(f):
            try:
                f.result()
            except _TRANSPORT_ERRORS as e:
                wrapped = GremlinConnectionError(_CONNECTION_ERROR_MSG)
                wrapped.__cause__ = e
                future.set_exception(wrapped)
                self._pool.put_nowait(self)
            except Exception as e:
                future.set_exception(e)
                self._pool.put_nowait(self)
            else:
                # Start receive task
                done = self._executor.submit(self._receive)
                self._result_set.done = done
                future.set_result(self._result_set)

        future_write.add_done_callback(cb)
        return future

    def _receive(self):
        try:
            # Surface error responses whose Content-Type doesn't match the
            # configured response serializer (e.g. plain-text 5xx).
            status = getattr(self._transport, 'status_code', None)
            response_content_type = str(self._response_serializer.version, encoding='utf-8')
            if status is not None and status >= 400:
                content_type = getattr(self._transport, 'content_type', '')
                if response_content_type not in content_type:
                    body = self._transport.read_body().decode('utf-8', errors='replace')
                    raise GremlinServerError({
                        'code': status,
                        'message': body,
                        'exception': ''
                    })

            # 204 No Content
            if status == 204:
                return

            # Stream items through the configured response serializer; it
            # raises GremlinServerError if the response ends with a non-success status.
            stream = self._transport.get_stream()
            for obj in self._response_serializer.deserialize_response_stream(stream):
                self._result_set.stream.put_nowait(obj)
        except _TRANSPORT_ERRORS as err:
            # Evict the dead connection from aiohttp's internal pool, ensuring subsequent
            # requests get a fresh connection.
            self._transport.evict_response()
            msg = 'Server returned an empty response body' if isinstance(err, asyncio.IncompleteReadError) and not err.partial else _CONNECTION_ERROR_MSG
            raise GremlinConnectionError(msg) from err
        except Exception:
            self._transport.evict_response()
            raise
        finally:
            self._pool.put_nowait(self)

    def __add_header(self, key, value):
        if self._headers is None:
            self._headers = dict()
        # Headers may be a list of pairs
        if isinstance(self._headers, list):
            for pair in self._headers:
                if pair[0] == key:
                    self._headers.remove(pair)
            self._headers.append((key, value))
        else:
            self._headers[key] = value

File: gremlin-js/gremlin-javascript/lib/driver/connection.ts
/*
 *  Licensed to the Apache Software Foundation (ASF) under one
 *  or more contributor license agreements.  See the NOTICE file
 *  distributed with this work for additional information
 *  regarding copyright ownership.  The ASF licenses this file
 *  to you under the Apache License, Version 2.0 (the
 *  "License"); you may not use this file except in compliance
 *  with the License.  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing,
 *  software distributed under the License is distributed on an
 *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *  KIND, either express or implied.  See the License for the
 *  specific language governing permissions and limitations
 *  under the License.
 */

/**
 * @author Jorge Bay Gondra
 */

import { Buffer } from 'buffer';
import { EventEmitter } from 'eventemitter3';
import type { Dispatcher } from 'undici';
import ioc, { createPreciseReader } from '../structure/io/binary/GraphBinary.js';
import GraphBinaryReader from '../structure/io/binary/internals/GraphBinaryReader.js';
import StreamReader from '../structure/io/binary/internals/StreamReader.js';
import * as utils from '../utils.js';
import ResultSet from './result-set.js';
import {RequestMessage} from "./request-message.js";
import { HttpRequest, RequestInterceptor } from './http-request.js';
import ResponseError from './response-error.js';
import { Traverser } from '../process/traversal.js';
import { buildDispatcher } from './dispatcher.js';
import { Logger, LoggerCallback, normalizeLogger } from './logger.js';

const responseStatusCode = {
  success: 200,
  noContent: 204,
  partialContent: 206,
};

/**
 * Selects the content encoding requested from, and decoded for, the server. `'deflate'` (the
 * default) sends an `Accept-Encoding: deflate` header and decodes deflate responses; `'none'`
 * turns compression off.
 */
export type Compression = 'none' | 'deflate';

export type ConnectionOptions = {
  preciseNumbers?: boolean;
  pdtRegistry?: any;
  reader?: any;
  traversalSource?: string;
  enableUserAgentOnConnect?: boolean;
  /** Maximum number of concurrent connections per origin. Defaults to 128. */
  maxConnections?: number;
  /** Idle-read (body) timeout in milliseconds, applied to the default dispatcher. */
  readTimeoutMillis?: number;
  /** Maximum size of the response headers in bytes, applied to the default dispatcher. */
  maxResponseHeaderBytes?: number;
  /**
   * Idle time in milliseconds before TCP keep-alive probes begin on a connection. Defaults to
   * 30000 (30s) when unset. Set to `0` to disable keep-alive entirely.
   */
  keepAliveTimeMillis?: number;
  /** HTTP proxy URI. When set, requests are routed through an undici `ProxyAgent`. */
  proxy?: string;
  /** Response compression codec. Defaults to `'deflate'` (on). */
  compression?: Compression;
  /** Connection-level default that fills a request's `batchSize` when it is left unset. Defaults to 64. */
  batchSize?: number;
  /** Connection-level default for `bulkResults`, applied to every request unless overridden per-request. Defaults to `false`. */
  bulkResults?: boolean;
  /** An optional logger (a logger object or a callback). Logging is disabled when unset. */
  logger?: Logger;
  interceptors?: RequestInterceptor | RequestInterceptor[];
  /** An optional auth interceptor. As a convenience, this is always appended to the end of the
   *  interceptor list so it runs last, after any user interceptors have modified the request. */
  auth?: RequestInterceptor;
};

/** The default per-request batch size used when neither the request nor the connection sets one. */
export const DEFAULT_BATCH_SIZE = 64;

/**
 * Represents a single connection to a Gremlin Server.
 */
export default class Connection extends EventEmitter {
  private readonly _reader: any;

  isOpen = true;
  traversalSource: string;

  private readonly _enableUserAgentOnConnect: boolean;
  private readonly _interceptors: RequestInterceptor[];
  private readonly _dispatcher: Dispatcher | undefined;
  private readonly _compression: Compression;
  private readonly _batchSize: number;
  private readonly _bulkResults: boolean;
  private readonly _log: LoggerCallback;

  /**
   * Creates a new instance of {@link Connection}.
   * @param {String} url The resource uri.
   * @param {ConnectionOptions} [options] The connection options.
   */
  constructor(
    readonly url: string,
    readonly options: ConnectionOptions = {},
  ) {
    super();

    this._reader = options.reader || (options.preciseNumbers === true ? createPreciseReader() : new GraphBinaryReader(ioc));
    if (options.pdtRegistry) {
      this._reader.pdtRegistry = options.pdtRegistry;
    }
    this.traversalSource = options.traversalSource || 'g';
    this._enableUserAgentOnConnect = options.enableUserAgentOnConnect !== false;
    this._log = normalizeLogger(options.logger);

    if (options.compression === undefined) {
      this._compression = 'deflate';
    } else if (options.compression === 'none' || options.compression === 'deflate') {
      this._compression = options.compression;
    } else {
      throw new TypeError(`compression must be 'none' or 'deflate'`);
    }

    this._batchSize = options.batchSize ?? DEFAULT_BATCH_SIZE;
    this._bulkResults = options.bulkResults ?? false;

    // The driver builds and owns the undici dispatcher from the discrete options.
    // TLS is configured through the Node/undici runtime (e.g. NODE_EXTRA_CA_CERTS),
    // not through a driver option.
    this._dispatcher = buildDispatcher({
      maxConnections: options.maxConnections,
      readTimeoutMillis: options.readTimeoutMillis,
      maxResponseHeaderBytes: options.maxResponseHeaderBytes,
      keepAliveTimeMillis: options.keepAliveTimeMillis,
      proxy: options.proxy,
    });

    const interceptors = options.interceptors;
    if (typeof interceptors === 'function') {
      this._interceptors = [interceptors];
    } else if (Array.isArray(interceptors)) {
      this._interceptors = [...interceptors];
    } else if (interceptors === undefined || interceptors === null) {
      this._interceptors = [];
    } else {
      throw new TypeError('interceptors must be a function, array, or undefined');
    }

    // Auth interceptor is always last so it runs after user interceptors
    if (options.auth) {
      this._interceptors.push(options.auth);
    }

    this._log('debug', `Connection created for ${this.url}`);
  }

  /**
   * The connection-level default batch size, used to fill a request's `batchSize` when unset.
   */
  get batchSize(): number {
    return this._batchSize;
  }

  /**
   * The connection-level default for `bulkResults`, applied to every request unless overridden
   * per-request. Defaults to `false`.
   */
  get bulkResults(): boolean {
    return this._bulkResults;
  }

  /**
   * Opens the connection, if its not already opened.
   * @returns {Promise}
   */
  async open() {
    // No-op for HTTP connections
    return Promise.resolve();
  }

  /**
   * Send a request and buffer the entire response. Returns a Promise<ResultSet>.
   */
  async submit(request: RequestMessage) {
    const response = await this.#makeHttpRequest(request);
    return this.#handleResponse(response);
  }

  /**
   * Send a request and stream the response incrementally.
   * Returns an AsyncGenerator that yields deserialized result items.
   * For bulked responses, yields Traverser objects.
   *
   * In the GraphBinary v4 streaming protocol, the server sends the status after all
   * result data. If the server encounters an error mid-traversal, values yielded before
   * the error are valid partial results. A ResponseError is thrown after the last value
   * has been yielded.
   *
   * @param {RequestMessage} request
   * @returns {AsyncGenerator<any>}
   */
  async *stream(request: RequestMessage): AsyncGenerator<any> {
    const abortController = new AbortController();

    let response: Response;
    try {
      response = await this.#makeHttpRequest(request, abortController.signal);
    } catch (e: any) {
      throw new Error(`Stream request failed: ${e.message}`, { cause: e });
    }

    if (!response.ok) {
      // For error responses, buffer and parse the error body
      const buffer = Buffer.from(await response.arrayBuffer());
      const errorMessage = `Server returned HTTP ${response.status}: ${response.statusText}`;
      const contentType = response.headers.get("Content-Type");
      const reader = this.#getReaderForContentType(contentType);

      try {
        if (reader) {
          const deserialized = await reader.readResponse(buffer);
          throw new ResponseError(errorMessage, {
            code: deserialized.status.code,
            message: deserialized.status.message || response.statusText,
            exception: deserialized.status.exception,
          });
        } else if (contentType === 'application/json') {
          const errorBody = JSON.parse(buffer.toString());
          throw new ResponseError(errorMessage, {
            code: response.status,
            message: errorBody.message || errorBody.error || response.statusText,
          });
        }
      } catch (err) {
        if (err instanceof ResponseError) throw err;
      }

      throw new ResponseError(errorMessage, {
        code: response.status,
        message: response.statusText,
      });
    }

    if (!response.body) {
      // 204 No Content - nothing to yield
      return;
    }

    const streamReader = StreamReader.fromReadableStream(response.body);

    let completed = false;
    try {
      yield* this._reader.readResponseStream(streamReader);
      completed = true;
    } finally {
      if (!completed) {
        // Consumer broke out early or an error occurred - abort to release the connection
        abortController.abort();
      }
    }
  }

  #getReaderForContentType(contentType: string | null) {
    if (!contentType) {
      return this._reader;
    }

    if (contentType === this._reader.mimeType) {
      return this._reader;
    }

    return null;
  }

  async #makeHttpRequest(request: RequestMessage, signal?: AbortSignal): Promise<Response> {
    const headers: Record<string, string> = {
      'Accept': this._reader.mimeType,
    };

    if (this._compression === 'deflate') {
      headers['Accept-Encoding'] = 'deflate';
    }

    if (this._enableUserAgentOnConnect) {
      const userAgent = await utils.getUserAgent();
      if (userAgent !== undefined) {
        headers[utils.getUserAgentHeader()] = userAgent;
      }
    }

    const httpRequest = new HttpRequest('POST', this.url, headers, request);

    // Promote transactionId to HTTP header before interceptors run.
    // The field remains in the serialized body as well (dual transmission
    // per the HTTP transaction protocol specification).
    if (request instanceof RequestMessage) {
      const fields = request.getFields();
      if (fields.has('transactionId')) {
        httpRequest.headers['X-Transaction-Id'] = fields.get('transactionId');
      }
    }

    for (let i = 0; i < this._interceptors.length; i++) {
      try {
        await this._interceptors[i](httpRequest);
      } catch (e: any) {
        throw new Error(`Request interceptor at index ${i} failed: ${e.message}`, { cause: e });
      }
    }

    // Auto-serialize body to JSON after interceptors run (idempotent if already serialized)
    httpRequest.serializeBody();

    this._log('debug', `Sending ${httpRequest.method} request to ${httpRequest.url}`);

    return fetch(httpRequest.url, {
      method: httpRequest.method,
      headers: httpRequest.headers,
      body: httpRequest.body,
      signal,
      // Node only: the undici dispatcher carries the connection-pool options. In the browser
      // it is undefined and the field is omitted, letting the user agent manage the transport.
      ...(this._dispatcher ? { dispatcher: this._dispatcher } : {}),
    } as RequestInit);
  }

  async #handleResponse(response: Response) {
    const contentType = response.headers.get("Content-Type");
    const buffer = Buffer.from(await response.arrayBuffer());
    const reader = this.#getReaderForContentType(contentType);

    if (!response.ok) {
      const errorMessage = `Server returned HTTP ${response.status}: ${response.statusText}`;

      try {
        if (reader) {
          const deserialized = await reader.readResponse(buffer);
          throw new ResponseError(errorMessage, {
            code: deserialized.status.code,
            message: deserialized.status.message || response.statusText,
            exception: deserialized.status.exception,
          });
        } else if (contentType === 'application/json') {
          const errorBody = JSON.parse(buffer.toString());
          throw new ResponseError(errorMessage, {
            code: response.status,
            message: errorBody.message || errorBody.error || response.statusText,
          });
        }
      } catch (err) {
        if (err instanceof ResponseError) {
          throw err;
        }
      }
      
      throw new ResponseError(errorMessage, {
        code: response.status,
        message: response.statusText,
      });
    }

    if (!reader) {
      throw new Error(`Response Content-Type '${contentType}' does not match the configured reader (expected '${this._reader.mimeType}')`);
    }

    const deserialized = await reader.readResponse(buffer);
    
    if (deserialized.status.code && deserialized.status.code !== responseStatusCode.success && deserialized.status.code !== responseStatusCode.noContent && deserialized.status.code !== responseStatusCode.partialContent) {
      throw new ResponseError(
        `Server error: ${deserialized.status.message || 'Unknown error'} (${deserialized.status.code})`,
        {
          code: deserialized.status.code,
          message: deserialized.status.message || '',
          exception: deserialized.status.exception,
        }
      );
    }

    return new ResultSet(
      deserialized.result.bulked
        ? deserialized.result.data.map((item: { v: any; bulk: number }) => new Traverser(item.v, item.bulk))
        : deserialized.result.data,
      new Map(),
    );
  }

  /**
   * Closes the Connection.
   * @return {Promise}
   */
  async close() {
    this._log('debug', `Closing connection to ${this.url}`);
    await this._dispatcher?.close();
  }
}
